pairtools/pairtools_dedup.py

```diff
@@ -274,14 +450,209 @@ def fetchadd(key, mydict):
+    dups = df["clusterid"].duplicated()
+    df["duplicate"] = False
+    if keep_parent_read_id:
+        df["parent_readID"] = df["clusterid"].map(
+            df[~dups].set_index("clusterid")["readID"]
+        )
+        unmapped["parent_readID"] = ""
+    unmapped["duplicate"] = False
+
+    df.iloc[dups, df.columns.get_loc("duplicate")] = True
+    return pd.concat([unmapped, df.drop(["clusterid"], axis=1)]).reset_index(drop=True)
+
+
+def _dedup_by_chunk(
+    in_stream,
+    colnames,
+    method,
+    chunksize,
+    mark_dups,
+    max_mismatch,
+    extra_col_pairs,
+    save_parent_id,
+    comment_char,
+    backend,
+    n_proc,
+):
+    dfs = pd.read_table(
+        in_stream, comment=comment_char, names=colnames, chunksize=chunksize
+    )
+
+    old_nodups = pd.DataFrame([])
+    old_i = 0
+    for df in dfs:
+        marked = dedup_chunk(
+            pd.concat([old_nodups, df], axis=0, ignore_index=True).reset_index(
+                drop=True
+            ),
+            r=max_mismatch,
+            method=method,
+            keep_parent_read_id=save_parent_id,
+            extra_col_pairs=extra_col_pairs,
+            backend=backend,
+            n_proc=n_proc,
+        )
+        marked = marked.iloc[old_i:, :].reset_index(drop=True)
+        if mark_dups:
+            marked.iloc[marked["duplicate"], marked.columns.get_loc("pair_type")] = "DD"
+        nodups = marked[~marked["duplicate"]]
+
+        nodups = nodups[colnames]
+        i = max(nodups.shape[0] // 100, 100)
+        old_nodups = nodups.iloc[-i:].reset_index(drop=True)
+        old_i = i
+        yield marked
+
+
+def streaming_dedup_by_chunk(
+    in_stream,
+    colnames,
+    chunksize,
+    method,
+    mark_dups,
+    max_mismatch,
+    extra_col_pairs,
+    unmapped_chrom,
+    comment_char,
+    outstream,
+    outstream_dups,
+    outstream_unmapped,
+    save_parent_id,
+    out_stat,
+    backend,
+    n_proc,
+):
+    deduped_chunks = _dedup_by_chunk(
+        in_stream=in_stream,
+        colnames=colnames,
+        method=method,
+        chunksize=chunksize,
+        mark_dups=mark_dups,
+        max_mismatch=max_mismatch,
+        extra_col_pairs=extra_col_pairs,
+        save_parent_id=save_parent_id,
+        comment_char=comment_char,
+        backend=backend,
+        n_proc=n_proc,
+    )
+    t0 = time.time()
+    N = 0
+    for chunk in deduped_chunks:
+        N += chunk.shape[0]
+        if out_stat is not None:
+            out_stat.add_pairs_from_dataframe(chunk, unmapped_chrom=unmapped_chrom)
+        mapped = np.logical_and(
+            (chunk["chrom1"] != unmapped_chrom), (chunk["chrom2"] != unmapped_chrom)
+        )
+        duplicates = chunk["duplicate"]
+        chunk = chunk.drop(columns=["duplicate"])
+        if outstream_dups:
+            chunk[mapped & duplicates].to_csv(
+                outstream_dups, index=False, header=False, sep="\t"
+            )
+            if save_parent_id:
```

Comment:
Seems like either indent or order of filtering/writing is wrong for outstream_dups here.

Reply:
Yeah, probably this should un-indented, just in case...